client-side-taxonomy-list/client-side-taxonomy-list.php

```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * Plugin Name:       Client Side Taxonomy List
+ * Description:       Display a list of terms in a selected taxonomy and
+ *                    adds Interactivity API directives to them to enable
+ *                    client-side navigation.
+ * Version:           0.1.0
+ * Requires at least: 6.1
+ * Requires PHP:      7.0
+ * Author:            The WordPress Contributors
+ * License:           GPL-2.0-or-later
+ * License URI:       https://www.gnu.org/licenses/gpl-2.0.html
+ * Text Domain:       client-side-taxonomy-list
+ *
+ * @package           wpcomsp
+ *
+ */
+
+/**
+ * Renders the `core/categories` block on server.
+ *
+ * @since 5.0.0
+ *
+ * @param array $attributes The block attributes.
+ *
+ * @return string Returns the categories list/dropdown markup.
+ */
+function render_block_video_categories( $attributes ) {
+	static $block_id = 0;
+	++$block_id;
+
+	if (empty( $attributes['selectedTaxonomy']) ) {
+		return '';
+	}
+
+	$args = array(
+		'echo'         => false,
+		'hierarchical' => ! empty( $attributes['showHierarchy'] ),
+		'orderby'      => 'name',
+		'show_count'   => ! empty( $attributes['showPostCounts'] ),
+		'title_li'     => '',
+		'hide_empty'   => empty( $attributes['showEmpty'] ),
+		'taxonomy' 	   => $attributes['selectedTaxonomy'],
+	);
+	if ( ! empty( $attributes['showOnlyTopLevel'] ) && $attributes['showOnlyTopLevel'] ) {
+		$args['parent'] = 0;
+	}
+
+	if ( ! empty( $attributes['displayAsDropdown'] ) ) {
+		$id                       = 'client-side-taxonomy-list-' . $block_id;
+		$args['id']               = $id;
+		$args['show_option_none'] = __( 'Select Video Category' );
+		$show_label               = empty( $attributes['showLabel'] ) ? ' screen-reader-text' : '';
+		$default_label            = __( 'Video Categories' );
+		$label_text               = ! empty( $attributes['label'] ) ? $attributes['label'] : $default_label;
+		$wrapper_markup           = '<div %1$s><label class="client-side-taxonomy-list__label' . $show_label . '" for="' . esc_attr( $id ) . '">' . $label_text . '</label>%2$s</div>';
+		$items_markup             = wp_dropdown_categories( $args );
+		$type                     = 'dropdown';
+
+		if ( ! is_admin() ) {
+			// Inject the dropdown script immediately after the select dropdown.
+			$items_markup = preg_replace(
+				'#(?<=</select>)#',
+				build_dropdown_script_block_core_categories( $id ),
+				$items_markup,
+				1
+			);
+		}
+	} else {
+		$wrapper_markup = '<ul %1$s>%2$s</ul>';
+		$items_markup   = wp_list_categories( $args );
+		$type           = 'list';
+	}
+
+	$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => "client-side-taxonomy-list-{$type}" ) );
+
+	$p = new WP_HTML_Tag_Processor( $items_markup );
+
+	while ( $p->next_tag( 'a' ) ) {
+		$p->set_attribute( 'data-wp-on--click', 'core/query::actions.navigate' );
+	}
+	$processed_items_markup = $p->get_updated_html();
```

Comment:
This is the part that adds the Interactivity API's `data-wp-on--click` directive in order to opt the block into client-side navigation, when enabled on the surrounding query block.

We could consider removing it and leaving it up to the developer to add it via the `render_block` filter, as we've done [here](https://github.com/a8cteam51/mike-flanagan-film/blob/7333087e42df34db37932057c437f5a9497b2f73/themes/mike-flanagan-film/functions.php#L61-L85) for the Categories block.

IMO, that'd be reasonable, as it's a simple enough change that might not always be needed. There'd still be value in having a Taxonomies block that works with any taxonomy, rather than just categories, which is what Core's Categories List block currently offers 🙂

Reply:
(One argument for removing the client-side navigation-specific code could be that we'd preserve closer parity to Core's Categories List block, on which this block is based; which might make it easier to eventually swap out this block for Core's, [whenever it finally supports custom taxonomies](https://github.com/WordPress/gutenberg/issues/33302).)

Reply:
Thanks for the feedback! All of this makes sense to me. I've changed the block to remove the directives 👍